### PR TITLE
Add tests to Trainer

### PR DESCRIPTION
**Pull request by sgugger**
This PR moves the tests of the various `data_collator` in `test_data_collator.py` and adds tests of the Trainer on a simple regression problem. While testing, a few problems were uncovered:
- The number of epochs is documented as a float but used as an int, fixed the documentation.
- There was one more step done than specified by the argument `max_steps`.
- The evaluation loss was wrong whenever the evaluation dataset length is not a round multiple of the batch size.

Those three things are also fixed in the PR. With the regression infrastructure, we can add more tests (for custom data collator, optimizers, schedulers etc...) since each training is fast. Will do in follow-up PRs as this one was starting to be of a decent size already.

### CODE REVIEW

**Comment by julien-c** [comment]
Re. the eval loss, did you also run [test_trainer_distributed.py](https://github.com/huggingface/transformers/blob/master/tests/test_trainer_distributed.py) on a multi-gpu machine?

**Comment by sgugger** [comment]
No, I don't have a multi-GPU machine setup. It does not seem like this test uses the eval_loss anywhere, it only computes a metric.

**Comment by julien-c** [comment]
> No, I don't have a multi-GPU machine setup.

You can use the office machines! Not necessarily related to this PR but to keep in mind to run this test once in a while

File: tests/test_trainer.py [hunk 1]
```diff
@@ -10,149 +12,39 @@
...
-
-        example = [torch.randint(5, [5])]
-        with self.assertRaises(ValueError):
-            # Expect error due to odd sequence length
-            data_collator(example)
+    def __len__(self):
+        return self.length
+
+    def __getitem__(self, i):
+        # Workaround the fact the default data collator wants tensors of ints except for the labels.
```
Comment (sgugger) [inline tests/test_trainer.py:36]: Thinking about this more, the `default_data_collator` should accept float tensors as inputs for `input_embeds` at least, so I could use this here.

**Review by sshleifer — APPROVED** [review]
randomly read this, nice tests!

**Comment by sgugger** [comment]
> Not necessarily related to this PR but to keep in mind to run this test once in a while

It was indeed broken not due to this PR, fixed. Know how to run it periodically now :-)

**Pull request merged**